core/platform/bulk_email/mailchimp_bulk_email_services_test.py

```diff
@@ -0,0 +1,212 @@
+from mailchimp3 import mailchimpclient
+
+import python_utils
+
+
+class MailchimpServicesUnitTests(test_utils.GenericTestBase):
+    """Tests for mailchimp services."""
+
+    def setUp(self):
+        super(MailchimpServicesUnitTests, self).setUp()
+        self.user_email_1 = 'test1@example.com'
+        self.user_email_2 = 'test2@example.com'
+        self.user_email_3 = 'test3@example.com'
+
+    class MockMailchimpClass(python_utils.OBJECT):
+        """Class to mock Mailchimp class."""
+
+        update_call_data = {}
+
+        class MailchimpLists(python_utils.OBJECT):
+            """Class to mock Mailchimp lists object."""
+
+            class MailchimpMembers(python_utils.OBJECT):
+                """Class to mock Mailchimp members object."""
+
+                def __init__(self):
+                    self.update_data = None
+                    self.create_data = None
+                    # This user is mocked to be present in mailchimp DB, but
+                    # unsubscribed initially. Hash corresponds to
+                    # 'test1@example.com'.
+                    self.subscribed_user = 'aa99b351245441b8ca95d54a52d2998c'
+                    # This user is mocked to be present in mailchimp DB, but
+                    # subscribed initially. Hash corresponds to
+                    # 'test2@example.com'.
+                    self.unsubscribed_user = '43b05f394d5611c54a1a9e8e20baee21'
+                    self.deleted_user = None
+
+                def get(self, _list_id, subscriber_hash):
+                    """Mocks the get function of the mailchimp api.
+
+                    Args:
+                        _list_id: str. List Id of mailchimp list.
+                        subscriber_hash: str. Subscriber hash, which is an MD5
+                            hash of subscriber's email ID.
+
+                    Raises:
+                        MailchimpError. Error 404 or 401 to mock API server
+                            error.
+
+                    Returns:
+                        dict. The updated status dict for users.
+                    """
+                    if subscriber_hash == self.subscribed_user:
+                        return {
+                            'status': 'unsubscribed'
+                        }
+                    elif subscriber_hash == self.unsubscribed_user:
+                        return {
+                            'status': 'subscribed'
+                        }
+                    elif not self.subscribed_user:
+                        raise mailchimpclient.MailChimpError(
+                            {'status': 401, 'detail': 'Server Error'})
+                    else:
+                        raise mailchimpclient.MailChimpError({'status': 404})
+
+                def update(self, _list_id, _subscriber_hash, data):
+                    """Mocks the update function of the mailchimp api. This
+                    function just sets the payload data to a private variable
```

Comment:
Hm, why don't you do more of a functional mock here? I.e. maintain a dictionary or something that serves as the "database" and update it accordingly, so that create()/update()/etc. actually work as expected.

Reply:
Done